// @tacoinfra/harbinger-lib/build/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initOracleLib = void 0;
const node_fetch_1 = __importDefault(require("node-fetch"));
const loglevel_1 = require("loglevel");
const conseiljs_1 = require("conseiljs");
function initOracleLib(conseilLogLevel = 'error') {
    const logger = loglevel_1.getLogger('conseiljs');
    logger.setLevel(conseilLogLevel, false);
    conseiljs_1.registerLogger(logger);
    conseiljs_1.registerFetch(node_fetch_1.default);
}
exports.initOracleLib = initOracleLib;
var update_1 = require("./update");
Object.defineProperty(exports, "updateOracleFromCoinbase", { enumerable: true, get: function () { return update_1.default; } });
Object.defineProperty(exports, "updateOracleFromCoinbaseOnce", { enumerable: true, get: function () { return update_1.updateOracleFromCoinbaseOnce; } });
Object.defineProperty(exports, "updateOracleFromFeed", { enumerable: true, get: function () { return update_1.updateOracleFromFeed; } });
Object.defineProperty(exports, "updateOracleFromFeedOnce", { enumerable: true, get: function () { return update_1.updateOracleFromFeedOnce; } });
var deploy_1 = require("./deploy");
Object.defineProperty(exports, "deployNormalizer", { enumerable: true, get: function () { return deploy_1.deployNormalizer; } });
Object.defineProperty(exports, "deployOracle", { enumerable: true, get: function () { return deploy_1.deployOracle; } });
var push_1 = require("./push");
Object.defineProperty(exports, "pushOracleData", { enumerable: true, get: function () { return push_1.default; } });
var common_1 = require("./common");
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return common_1.LogLevel; } });
var get_1 = require("./get");
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return get_1.default; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "Utils", { enumerable: true, get: function () { return utils_1.default; } });
var asn1_1 = require("./asn1");
Object.defineProperty(exports, "ASN1", { enumerable: true, get: function () { return asn1_1.default; } });
var revoke_1 = require("./revoke");
Object.defineProperty(exports, "revokeOracle", { enumerable: true, get: function () { return revoke_1.default; } });
var prefixes_1 = require("./prefixes");
Object.defineProperty(exports, "Prefixes", { enumerable: true, get: function () { return prefixes_1.default; } });
var constants_1 = require("./constants");
Object.defineProperty(exports, "Constants", { enumerable: true, get: function () { return constants_1.default; } });
var operation_fee_estimator_1 = require("./operation-fee-estimator");
Object.defineProperty(exports, "OperationFeeEstimator", { enumerable: true, get: function () { return operation_fee_estimator_1.default; } });
