// conseiljs-softsigner/dist/KeyStoreUtils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const bip39 = __importStar(require("bip39"));
const conseiljs_1 = require("conseiljs");
const conseiljs_2 = require("conseiljs");
const CryptoUtils_1 = require("./utils/CryptoUtils");
var KeyStoreUtils;
(function (KeyStoreUtils) {
    function generateMnemonic(strength = 256) {
        return bip39.generateMnemonic(strength);
    }
    KeyStoreUtils.generateMnemonic = generateMnemonic;
    function generateIdentity(strength = 256, password = '', mnemonic) {
        return __awaiter(this, void 0, void 0, function* () {
            return restoreIdentityFromMnemonic((mnemonic || bip39.generateMnemonic(strength)), password);
        });
    }
    KeyStoreUtils.generateIdentity = generateIdentity;
    function restoreIdentityFromSecretKey(secretKey) {
        return __awaiter(this, void 0, void 0, function* () {
            const secretKeyBytes = conseiljs_2.TezosMessageUtils.writeKeyWithHint(secretKey, 'edsk');
            const keys = yield recoverKeys(secretKeyBytes);
            const publicKey = conseiljs_2.TezosMessageUtils.readKeyWithHint(keys.publicKey, 'edpk');
            const publicKeyHash = conseiljs_2.TezosMessageUtils.computeKeyHash(keys.publicKey, 'tz1');
            return { publicKey, secretKey, publicKeyHash, curve: conseiljs_1.KeyStoreCurve.ED25519, storeType: conseiljs_1.KeyStoreType.Mnemonic };
        });
    }
    KeyStoreUtils.restoreIdentityFromSecretKey = restoreIdentityFromSecretKey;
    function restoreIdentityFromMnemonic(mnemonic, password = '', pkh, derivationPath, validate = true) {
        return __awaiter(this, void 0, void 0, function* () {
            if (validate) {
                if (![12, 15, 18, 21, 24].includes(mnemonic.split(' ').length)) {
                    throw new Error('Invalid mnemonic length.');
                }
                if (!bip39.validateMnemonic(mnemonic)) {
                    throw new Error('The given mnemonic could not be validated.');
                }
            }
            const seed = (yield bip39.mnemonicToSeed(mnemonic, password)).slice(0, 32);
            const keys = yield generateKeys(seed);
            const secretKey = conseiljs_2.TezosMessageUtils.readKeyWithHint(keys.secretKey, 'edsk');
            const publicKey = conseiljs_2.TezosMessageUtils.readKeyWithHint(keys.publicKey, 'edpk');
            const publicKeyHash = conseiljs_2.TezosMessageUtils.computeKeyHash(keys.publicKey, 'tz1');
            if (!!pkh && publicKeyHash !== pkh) {
                throw new Error('The given mnemonic and passphrase do not correspond to the supplied public key hash');
            }
            return { publicKey, secretKey, publicKeyHash, curve: conseiljs_1.KeyStoreCurve.ED25519, storeType: conseiljs_1.KeyStoreType.Mnemonic, seed: mnemonic, derivationPath };
        });
    }
    KeyStoreUtils.restoreIdentityFromMnemonic = restoreIdentityFromMnemonic;
    function restoreIdentityFromFundraiser(mnemonic, email, password, pkh) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield restoreIdentityFromMnemonic(mnemonic, email + password, pkh);
        });
    }
    KeyStoreUtils.restoreIdentityFromFundraiser = restoreIdentityFromFundraiser;
    function generateKeys(seed) {
        return __awaiter(this, void 0, void 0, function* () {
            const keys = yield CryptoUtils_1.CryptoUtils.generateKeys(seed);
            return { publicKey: keys.publicKey, secretKey: keys.privateKey };
        });
    }
    KeyStoreUtils.generateKeys = generateKeys;
    function recoverKeys(secretKey) {
        return __awaiter(this, void 0, void 0, function* () {
            const keys = yield CryptoUtils_1.CryptoUtils.recoverPublicKey(secretKey);
            return { publicKey: keys.publicKey, secretKey: keys.privateKey };
        });
    }
    KeyStoreUtils.recoverKeys = recoverKeys;
    function decryptMessage(message, passphrase, salt) {
        return __awaiter(this, void 0, void 0, function* () {
            return CryptoUtils_1.CryptoUtils.decryptMessage(message, passphrase, salt);
        });
    }
    KeyStoreUtils.decryptMessage = decryptMessage;
    function encryptMessage(message, passphrase, salt) {
        return __awaiter(this, void 0, void 0, function* () {
            return CryptoUtils_1.CryptoUtils.encryptMessage(message, passphrase, salt);
        });
    }
    KeyStoreUtils.encryptMessage = encryptMessage;
    function checkTextSignature(signature, message, publicKey, prehash = false) {
        return __awaiter(this, void 0, void 0, function* () {
            let messageBytes;
            if (prehash) {
                messageBytes = conseiljs_2.TezosMessageUtils.simpleHash(Buffer.from(message, 'utf8'), 32);
            }
            else {
                messageBytes = Buffer.from(message, 'utf8');
            }
            const sig = conseiljs_2.TezosMessageUtils.writeSignatureWithHint(signature, 'edsig');
            const pk = conseiljs_2.TezosMessageUtils.writeKeyWithHint(publicKey, 'edpk');
            return yield CryptoUtils_1.CryptoUtils.checkSignature(sig, messageBytes, pk);
        });
    }
    KeyStoreUtils.checkTextSignature = checkTextSignature;
})(KeyStoreUtils = exports.KeyStoreUtils || (exports.KeyStoreUtils = {}));
